core/src/main/java/com/sequenceiq/cloudbreak/job/StackStatusCheckerJob.java

```diff
@@ -288,20 +288,25 @@ private boolean isCMRunning(ClusterApi connector) {
         return Sets.intersect(healthyHosts, unhealthyStoredHosts);
     }
 
-    private Set<InstanceMetaData> getFailedInstancesInstanceMetadata(Map<HostName, ClusterManagerState> hostStatuses,
+    private Map<InstanceMetaData, Optional<String>> getFailedInstancesInstanceMetadata(ExtendedHostStatuses hostStatuses,
             Set<InstanceMetaData> runningInstances) {
-        Set<String> failedHosts = hostStatuses.entrySet().stream()
-                .filter(e -> e.getValue().getClusterManagerStatus() == ClusterManagerStatus.UNHEALTHY)
-                .map(Entry::getKey)
-                .map(HostName::value)
-                .collect(toSet());
+        Map<String, Optional<String>> failedHosts = hostStatuses.getHostsHealth().entrySet().stream()
+                .filter(e -> !hostStatuses.isHostHealthy(e.getKey()))
+                .collect(Collectors.toMap(e -> e.getKey().value(), e -> Optional.ofNullable(hostStatuses.statusReasonForHost(e.getKey()))));
         Set<String> noReportHosts = runningInstances.stream()
-                .filter(i -> hostStatuses.get(hostName(i.getDiscoveryFQDN())) == null)
+                .filter(i -> hostStatuses.getHostsHealth().get(hostName(i.getDiscoveryFQDN())) == null)
                 .map(InstanceMetaData::getDiscoveryFQDN)
                 .collect(toSet());
         return runningInstances.stream()
-                .filter(i -> failedHosts.contains(i.getDiscoveryFQDN()) || noReportHosts.contains(i.getDiscoveryFQDN()))
-                .collect(toSet());
+                .filter(i -> failedHosts.containsKey(i.getDiscoveryFQDN()) || noReportHosts.contains(i.getDiscoveryFQDN()))
+                .collect(Collectors.toMap(imd -> imd, imd -> getReasonForFailedInstance(failedHosts, imd)));
+    }
+
+    private Optional<String> getReasonForFailedInstance(Map<String, Optional<String>> failedHosts, InstanceMetaData imd) {
+        if (failedHosts.containsKey(imd.getDiscoveryFQDN())) {
+            return failedHosts.get(imd.getDiscoveryFQDN());
```

Comment:
I think Map#get would return null if it's not presetn, in that case you could just use:
`Optional.ofNullable(failedHosts.get(fqdn)`

Reply:
yes, it will `null`, good catch, i will check all `get`s to make sure

Reply:
but there is an `if` to check if there is key for the given fqdn, so this should work